File: baekjoon/String/4949.cpp
// 백준 4949번 : 균형잡힌 세상
#include <iostream>
#include <stack>

using namespace std;

int main() {

    while (true) {
        string text = "";
        getline(cin, text);

        if (text == ".") {
            break;
        }

        stack<char> st;
        bool flag = true;
        for (int i = 0; i < text.length(); i++) {
            char ch = text[i];
            if (ch == '(' || ch == '[') {
                st.push(ch);
            } else if (ch == ')') {
                if (st.empty() || st.top() != '(') {
                    flag = false;
                    break;
                } else {
                    st.pop();
                }
            } else if (ch == ']') {
                if (st.empty() || st.top() != '[') {
                    flag = false;
                    break;
                } else {
                    st.pop();
                }
            }
        }
        if (flag && st.empty()) {
            cout << "yes" << '\n';
        } else {
            cout << "no" << '\n';
        }
    }
    return 0;
}

File: baekjoon/String/4949.py
# 백준 4949번 : 균형잡힌 세상
if __name__ == "__main__":
    while True:
        sen = input()
        if sen == '.':
            break
        
        st = []

        for ch in sen:
            if ch == '(' or ch == '[':
                st.append(ch)
            elif ch == ')':
                if len(st) != 0 and st[-1] == '(':
                    st.pop()
                else:
                    st.append(ch)
                    break
            elif ch == ']':
                if len(st) != 0 and st[-1] == '[':
                    st.pop()
                else:
                    st.append(ch)
                    break
        
        if len(st) == 0:
            print('yes')
        else:
            print('no')